Add host platform support to MD5 component

dwmw2 opened the pull request:
Use OpenSSL EVP interface for MD5 on host platform, enabling HMAC-MD5 testing. Verified with known test vector.

# What does this implement/fix?

<!-- Quick description and explanation of changes -->

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Developer breaking change (an API change that could break external components)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840
- [x] host

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

dwmw2 committed:
Add host platform support to MD5 component

Use OpenSSL EVP interface for MD5 on host platform, enabling
HMAC-MD5 testing. Verified with known test vector.

bdraco commented on esphome/components/md5/md5.cpp:
```diff
@@ -39,6 +39,22 @@ void MD5Digest::add(const uint8_t *data, size_t len) { br_md5_update(&this->ctx_
+
+void MD5Digest::calculate() {
+  unsigned int len = 16;
+  EVP_DigestFinal_ex(this->ctx_, this->digest_, &len);
```
SHA256 uses a `calculated_` flag to prevent double finalization as it would crash if called twice

bdraco commented on esphome/components/md5/md5.cpp:
```diff
@@ -39,6 +39,22 @@ void MD5Digest::add(const uint8_t *data, size_t len) { br_md5_update(&this->ctx_
 void MD5Digest::calculate() { br_md5_out(&this->ctx_, this->digest_); }
 #endif  // USE_RP2040
 
+#ifdef USE_HOST
```
There is a destructor for SHA256 that is a bit safer to ensure it doesn't leak  on failure

https://github.com/esphome/esphome/blob/51b187954a411378f3c151f3424a41e65b805960/esphome/components/sha256/sha256.cpp#L77

dwmw2 replied:
Huh, I thought I'd literally taken the MD5 one and changed the hash. Fixed now; sorry about that.

bdraco commented on esphome/components/md5/md5.cpp:
```diff
@@ -39,6 +39,22 @@ void MD5Digest::add(const uint8_t *data, size_t len) { br_md5_update(&this->ctx_
 
+#ifdef USE_HOST
+void MD5Digest::init() {
+  memset(this->digest_, 0, 16);
```
could be a bit safer like SHA256 in case of misuse
https://github.com/esphome/esphome/blob/51b187954a411378f3c151f3424a41e65b805960/esphome/components/sha256/sha256.cpp#L84

bdraco commented on esphome/components/md5/md5.cpp:
```diff
@@ -39,6 +39,22 @@ void MD5Digest::add(const uint8_t *data, size_t len) { br_md5_update(&this->ctx_
+  EVP_DigestInit_ex(this->ctx_, EVP_md5(), nullptr);
+}
+
+void MD5Digest::add(const uint8_t *data, size_t len) { EVP_DigestUpdate(this->ctx_, data, len); }
```
same here https://github.com/esphome/esphome/blob/51b187954a411378f3c151f3424a41e65b805960/esphome/components/sha256/sha256.cpp#L93

dwmw2 committed:
Add hmac-md5 test

dwmw2 committed:
Merge branch 'dev' of github.com:esphome/esphome into md5-host

dwmw2 committed:
Fix md5 USE_HOST implementation to match sha256 pattern

- Add proper destructor to free EVP_MD_CTX (fixes memory leak)
- Add calculated_ flag to prevent double-calculation
- Add lazy initialization in add() and calculate()
- Make init() reusable by cleaning up old context
- Move context cleanup from calculate() to destructor

dwmw2 commented on esphome/components/md5/__init__.py:
```diff
@@ -1,7 +1,17 @@
     cg.add_define("USE_MD5")
+
+    # Add OpenSSL library for host platform
+    if CORE.is_host:
+        if IS_MACOS:
+            # macOS needs special handling for Homebrew OpenSSL
+            cg.add_build_flag("-I/opt/homebrew/opt/openssl/include")
+            cg.add_build_flag("-L/opt/homebrew/opt/openssl/lib")
+        cg.add_build_flag("-lcrypto")
```
Tested via hmac-md5.

bdraco commented on esphome/components/md5/md5.h:
```diff
@@ -25,13 +25,35 @@
+#if defined(USE_LIBRETINY)
+#define MD5_CTX_TYPE LT_MD5_CTX_T
+#endif
+#endif
```
Why redefine again ?

dwmw2 committed:
Remove duplicate MD5_CTX_TYPE definitions in md5.h

bdraco reviewed (approved):
Thanks @dwmw2
